File: erickson/balanced-brackets/balanced-brackets.cpp
#include <iostream>
#include <stack>
#include <string>

int main(void)
{
    unsigned int n;
    std::cin >> n;
    for (auto t = 0; t != n; ++t) {
        std::string input;
        std::cin >> input;
        std::stack<unsigned char> stack;
        auto proceed = true;
        for (auto i = 0; i != input.size() && proceed; ++i) {
            switch (input[i]) {
                case '{':
                case '[':
                case '(':
                    stack.push(input[i]);
                    break;
                case '}':
                    if (stack.size() > 0 && stack.top() == '{')
                        stack.pop();
                    else
                        proceed = false;
                    break;
                case ']':
                    if (stack.size() > 0 && stack.top() == '[')
                        stack.pop();
                    else
                        proceed = false;
                    break;
                case ')':
                    if (stack.size() > 0 && stack.top() == '(')
                        stack.pop();
                    else
                        proceed = false;
                    break;
                default:
                    break;
            }
        }
        std::cout << (stack.size() == 0 && proceed ? "YES" : "NO") << std::endl;
    }
    return 0;
}

File: henrique/balanced-brackets/src/br/com/hcavalcante/BalancedBrackets.java
package br.com.hcavalcante;

import java.util.Scanner;
import java.util.Stack;

public class BalancedBrackets {

	public static void main(String[] args) {
		int i, j, t, strLength, closingBrackets;
		String str;

		Scanner input = new Scanner(System.in);
		t = input.nextInt();

		Stack<Character> openingBrackets;
		
		for (i = 0; i < t; i++) {
			openingBrackets = new Stack<Character>();
			closingBrackets = 0;

			str = input.next();
			strLength = str.length();

			for (j = 0; j < strLength; j++) {
				if (str.charAt(j) == '(' || str.charAt(j) == '[' || str.charAt(j) == '{') {
					openingBrackets.push(str.charAt(j));
				} else {
					closingBrackets++;
					if (!openingBrackets.isEmpty()) {
						if ((openingBrackets.peek().equals('(') && str.charAt(j) == ')')
								|| (openingBrackets.peek().equals('[') && str.charAt(j) == ']')
								|| (openingBrackets.peek().equals('{') && str.charAt(j) == '}')) {
							openingBrackets.pop();
							closingBrackets--;
						}
					}
				}
			}

			if (openingBrackets.isEmpty() && closingBrackets == 0) {
				System.out.println("YES");
			} else {
				System.out.println("NO");
			}

			openingBrackets.clear();
		}

		input.close();
	}
}